File: seeddao/initialize-dao.js
const anchor = require('@coral-xyz/anchor');
const { PublicKey, Keypair, SystemProgram } = require('@solana/web3.js');
const { TOKEN_PROGRAM_ID, createMint, getOrCreateAssociatedTokenAccount } = require('@solana/spl-token');

// SeedDAO initialization script
async function initializeDao() {
    // Configure the client to use devnet
    const provider = anchor.AnchorProvider.env();
    anchor.setProvider(provider);

    // Load the program
    const program = anchor.workspace.Seeddao;
    const wallet = provider.wallet;

    console.log('Initializing SeedDAO...');
    console.log('Program ID:', program.programId.toString());
    console.log('Wallet:', wallet.publicKey.toString());

    // DAO configuration
    const totalSupply = new anchor.BN(10_000_000 * 1e6); // 10M SDAO tokens
    const minProposalThreshold = new anchor.BN(1_000 * 1e6); // 1,000 SDAO to create proposal
    const votingPeriod = new anchor.BN(7 * 24 * 3600); // 7 days
    const executionDelay = new anchor.BN(2 * 24 * 3600); // 2 days
    const quorumThreshold = new anchor.BN(100_000 * 1e6); // 100,000 SDAO for quorum

    // Derive PDAs
    const [daoState] = PublicKey.findProgramAddressSync(
        [Buffer.from('dao_state')],
        program.programId
    );

    const [treasury] = PublicKey.findProgramAddressSync(
        [Buffer.from('treasury'), daoState.toBuffer()],
        program.programId
    );

    const [sdaoMint] = PublicKey.findProgramAddressSync(
        [Buffer.from('sdao_mint')],
        program.programId
    );

    // Mock USDC and SEEDS mints (replace with actual mint addresses)
    const usdcMint = new PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'); // USDC devnet
    const seedsMint = new PublicKey('Eoyy5BhjVsRUTiyHoNbM675PAZHdyX7qGr1yndZezYQG'); // Your SEEDS mint

    const [usdcVault] = PublicKey.findProgramAddressSync(
        [Buffer.from('usdc_vault'), treasury.toBuffer()],
        program.programId
    );

    const [seedsVault] = PublicKey.findProgramAddressSync(
        [Buffer.from('seeds_vault'), treasury.toBuffer()],
        program.programId
    );

    try {
        const tx = await program.methods
            .initializeDao(
                totalSupply,
                minProposalThreshold,
                votingPeriod,
                executionDelay,
                quorumThreshold
            )
            .accounts({
                daoState,
                treasury,
                sdaoMint,
                usdcVault,
                seedsVault,
                usdcMint,
                seedsMint,
                authority: wallet.publicKey,
                systemProgram: SystemProgram.programId,
                tokenProgram: TOKEN_PROGRAM_ID,
                rent: anchor.web3.SYSVAR_RENT_PUBKEY,
            })
            .rpc();

        console.log('✅ SeedDAO initialized successfully!');
        console.log('Transaction signature:', tx);
        console.log('DAO State:', daoState.toString());
        console.log('SDAO Mint:', sdaoMint.toString());
        console.log('Treasury:', treasury.toString());
        
        return {
            daoState,
            sdaoMint,
            treasury,
            signature: tx
        };

    } catch (error) {
        console.error('❌ Failed to initialize DAO:', error);
        throw error;
    }
}

// Run if called directly
if (require.main === module) {
    initializeDao()
        .then(() => process.exit(0))
        .catch((error) => {
            console.error(error);
            process.exit(1);
        });
}

module.exports = { initializeDao };

File: seeddao/programs/seeddao/src/instructions/initialize_dao.rs
use anchor_lang::prelude::*;
use anchor_spl::token::{Mint, Token, TokenAccount};
use crate::state::*;

#[derive(Accounts)]
pub struct InitializeDao<'info> {
    #[account(
        init,
        payer = authority,
        space = DaoState::LEN,
        seeds = [b"dao_state"],
        bump
    )]
    pub dao_state: Account<'info, DaoState>,

    #[account(
        init,
        payer = authority,
        space = Treasury::LEN,
        seeds = [b"treasury", dao_state.key().as_ref()],
        bump
    )]
    pub treasury: Account<'info, Treasury>,

    #[account(
        init,
        payer = authority,
        mint::decimals = 6,
        mint::authority = dao_state,
        seeds = [b"sdao_mint"],
        bump
    )]
    pub sdao_mint: Account<'info, Mint>,

    #[account(
        init,
        payer = authority,
        token::mint = usdc_mint,
        token::authority = treasury,
        seeds = [b"usdc_vault", treasury.key().as_ref()],
        bump
    )]
    pub usdc_vault: Account<'info, TokenAccount>,

    #[account(
        init,
        payer = authority,
        token::mint = seeds_mint,
        token::authority = treasury,
        seeds = [b"seeds_vault", treasury.key().as_ref()],
        bump
    )]
    pub seeds_vault: Account<'info, TokenAccount>,

    pub usdc_mint: Account<'info, Mint>,
    pub seeds_mint: Account<'info, Mint>,

    #[account(mut)]
    pub authority: Signer<'info>,
    pub system_program: Program<'info, System>,
    pub token_program: Program<'info, Token>,
    pub rent: Sysvar<'info, Rent>,
}

pub fn handler(
    ctx: Context<InitializeDao>,
    total_supply: u64,
    min_proposal_threshold: u64,
    voting_period: i64,
    execution_delay: i64,
    quorum_threshold: u64,
) -> Result<()> {
    let dao_state = &mut ctx.accounts.dao_state;
    let treasury = &mut ctx.accounts.treasury;

    // Initialize DAO state
    dao_state.authority = ctx.accounts.authority.key();
    dao_state.sdao_mint = ctx.accounts.sdao_mint.key();
    dao_state.treasury = ctx.accounts.treasury.key();
    dao_state.total_supply = total_supply;
    dao_state.total_staked = 0;
    dao_state.proposal_count = 0;
    dao_state.min_proposal_threshold = min_proposal_threshold;
    dao_state.voting_period = voting_period;
    dao_state.execution_delay = execution_delay;
    dao_state.quorum_threshold = quorum_threshold;
    dao_state.is_paused = false;
    dao_state.bump = ctx.bumps.dao_state;

    // Initialize treasury
    treasury.dao_state = dao_state.key();
    treasury.usdc_vault = ctx.accounts.usdc_vault.key();
    treasury.seeds_vault = ctx.accounts.seeds_vault.key();
    treasury.total_value = 0;
    treasury.bump = ctx.bumps.treasury;

    msg!("SeedDAO initialized with total supply: {}", total_supply);
    msg!("Min proposal threshold: {}", min_proposal_threshold);
    msg!("Voting period: {} seconds", voting_period);

    Ok(())
}
